File: core/quader_engine/src/piece.rs
use std::fmt::{Display, Formatter};
use serde::{Deserialize, Serialize};
use crate::cell_holder::CellType;
use crate::piece_points;
use crate::primitives::{Point, Rect, Color};
use crate::utils::calc_bounds;
use crate::wall_kick_data::{WallKickType};

#[derive(Debug, Clone, Copy, PartialEq, Serialize, Deserialize)]
pub enum PieceType {
    I, O, T, L, J, S, Z, Pixel
}

impl Display for PieceType {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        let str = match self {
            PieceType::I => "I",
            PieceType::O => "O",
            PieceType::T => "T",
            PieceType::L => "L",
            PieceType::J => "J",
            PieceType::S => "S",
            PieceType::Z => "Z",
            PieceType::Pixel => "Pixel",
        };

        write!(f, "{}", str)
    }
}

#[derive(Debug, Clone, Copy, Eq, PartialEq, Serialize, Deserialize)]
pub enum RotationState {
    Initial = 0, Clockwise = 1, Deg180 = 2, CounterClockwise = 3
}

#[derive(Debug, Clone, Copy, Eq, PartialEq, Serialize, Deserialize)]
pub enum RotationDirection {
    Clockwise, CounterClockwise, Deg180
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize)]
pub enum RotationMove {
    InitToRight,
    RightToInit,

    RightToDeg180,
    Deg180ToRight,

    Deg180ToLeft,
    LeftToDeg180,

    LeftToInit,
    InitToLeft,

    InitToDeg180,
    Deg180ToInit
}

#[derive(Debug, Clone, Copy, PartialEq, Serialize, Deserialize)]
pub enum OffsetType {
    Cell,
    BetweenCells
}

fn get_points_for_piece(piece_type: PieceType, state: RotationState) -> &'static [Point] {
    match piece_type {
        PieceType::I => {
            match state {
                RotationState::Initial => &piece_points::piece_i::INIT_POS,
                RotationState::Clockwise => &piece_points::piece_i::RIGHT_POS,
                RotationState::CounterClockwise => &piece_points::piece_i::LEFT_POS,
                RotationState::Deg180 => &piece_points::piece_i::DEG180_POS,
            }
        },
        PieceType::O => {
            match state {
                RotationState::Initial => &piece_points::piece_o::INIT_POS,
                RotationState::Clockwise => &piece_points::piece_o::RIGHT_POS,
                RotationState::CounterClockwise => &piece_points::piece_o::LEFT_POS,
                RotationState::Deg180 => &piece_points::piece_o::DEG180_POS,
            }
        },
        PieceType::T =>
            match state {
                RotationState::Initial => &piece_points::piece_t::INIT_POS,
                RotationState::Clockwise => &piece_points::piece_t::RIGHT_POS,
                RotationState::CounterClockwise => &piece_points::piece_t::LEFT_POS,
                RotationState::Deg180 => &piece_points::piece_t::DEG180_POS,
            },
        PieceType::L => {
            match state {
                RotationState::Initial => &piece_points::piece_l::INIT_POS,
                RotationState::Clockwise => &piece_points::piece_l::RIGHT_POS,
                RotationState::CounterClockwise => &piece_points::piece_l::LEFT_POS,
                RotationState::Deg180 => &piece_points::piece_l::DEG180_POS,
            }
        },
        PieceType::J => {
            match state {
                RotationState::Initial => &piece_points::piece_j::INIT_POS,
                RotationState::Clockwise => &piece_points::piece_j::RIGHT_POS,
                RotationState::CounterClockwise => &piece_points::piece_j::LEFT_POS,
                RotationState::Deg180 => &piece_points::piece_j::DEG180_POS,
            }
        },
        PieceType::S => {
            match state {
                RotationState::Initial => &piece_points::piece_s::INIT_POS,
                RotationState::Clockwise => &piece_points::piece_s::RIGHT_POS,
                RotationState::CounterClockwise => &piece_points::piece_s::LEFT_POS,
                RotationState::Deg180 => &piece_points::piece_s::DEG180_POS,
            }
        },
        PieceType::Z => {
            match state {
                RotationState::Initial => &piece_points::piece_z::INIT_POS,
                RotationState::Clockwise => &piece_points::piece_z::RIGHT_POS,
                RotationState::CounterClockwise => &piece_points::piece_z::LEFT_POS,
                RotationState::Deg180 => &piece_points::piece_z::DEG180_POS,
            }
        },
        _ => panic!("invalid piece type")
    }
}

#[derive(Debug)]
pub struct WallKickCheckParams<'a> {
    pub tests: &'a [Point],
    pub expected_pos: &'a [Point]
}

#[derive(Debug)]
pub struct Piece {
    piece_type: PieceType,
    board_cell_type: CellType,
    offset_type: OffsetType,
    bounds: Rect,
    x: u32,
    y: u32,
    current_rotation: RotationState,
    wall_kick_type: WallKickType
}

impl Piece {
    pub fn new(piece_type: PieceType) -> Self {
        let wall_kick_type = match piece_type {
            PieceType::I => WallKickType::PieceI,
            PieceType::O => WallKickType::PieceO,
            _ => WallKickType::Default
        };

        let offset_type = match piece_type {
            PieceType::I | PieceType::O => OffsetType::BetweenCells,
            _ => OffsetType::Cell
        };

        let board_cell_type = crate::utils::piece_type_to_cell_type(piece_type);
        let init_pos = get_points_for_piece(piece_type, RotationState::Initial);

        Piece {
            bounds: calc_bounds(init_pos, 0, 0),
            piece_type,
            board_cell_type,
            offset_type,
            x: 0,
            y: 0,
            current_rotation: RotationState::Initial,
            wall_kick_type,
        }
    }

    pub fn reset(&mut self) {
        self.current_rotation = RotationState::Initial;
        self.bounds = self.calc_bounds();
    }

    pub fn get_positions(&self) -> &[Point] {
        get_points_for_piece(self.piece_type, self.current_rotation)
    }

    pub fn get_type(&self) -> PieceType {
        self.piece_type
    }

    pub fn get_cell_type(&self) -> CellType { self.board_cell_type }

    pub fn set_x(&mut self, x: u32) {
        self.x = x;
        self.bounds = self.calc_bounds();
    }

    pub fn set_y(&mut self, y: u32) {
        self.y = y;
        self.bounds = self.calc_bounds();
    }

    pub fn get_x(&self) -> u32 {
        self.x
    }

    pub fn get_y(&self) -> u32 {
        self.y
    }

    pub fn move_left(&mut self) {
        self.set_x(self.x - 1);
    }

    pub fn move_right(&mut self) {
        self.set_x(self.x + 1);
    }

    pub fn move_down(&mut self) {
        self.set_y(self.y + 1);
    }

    pub fn get_offset_type(&self) -> OffsetType {
        self.offset_type
    }

    pub fn get_current_pos(&self) -> &[Point] {
        get_points_for_piece(self.piece_type, self.current_rotation)
    }

    pub fn get_wall_kick_type(&self) -> &WallKickType {
        &self.wall_kick_type
    }

    pub fn rotate(&mut self, rotation: RotationDirection, x_offset: i32, y_offset: i32) {
        self.rotate_simple(rotation);

        let ix = self.x as i32 + x_offset;
        let iy = self.y as i32 + y_offset;

        self.x = ix as u32;
        self.y = iy as u32;

        self.bounds = self.calc_bounds();
    }

    pub fn get_bounds(&self) -> Rect {
        self.bounds
    }

    pub fn get_color(&self) -> Color {
        match self.piece_type {
            PieceType::I => *Color::PIECE_I,
            PieceType::O => *Color::PIECE_O,
            PieceType::T => *Color::PIECE_T,
            PieceType::L => *Color::PIECE_L,
            PieceType::J => *Color::PIECE_J,
            PieceType::S => *Color::PIECE_S,
            PieceType::Z => *Color::PIECE_Z,
            PieceType::Pixel => *Color::PIECE_GARBAGE,
        }
    }

    fn calc_bounds(&self) -> Rect {
        calc_bounds(self.get_current_pos(), self.x as i32, self.y as i32)
    }

    fn rotate_simple(&mut self, rotation: RotationDirection) {
        match rotation {
            RotationDirection::Clockwise => self.rotate_right(),
            RotationDirection::CounterClockwise => self.rotate_left(),
            RotationDirection::Deg180 => { self.rotate_right(); self.rotate_right(); }
        }

        self.bounds = self.calc_bounds();
    }

    fn rotate_right(&mut self) {
        self.current_rotation = match self.current_rotation {
            RotationState::Initial => RotationState::Clockwise,
            RotationState::Clockwise => RotationState::Deg180,
            RotationState::Deg180 => RotationState::CounterClockwise,
            RotationState::CounterClockwise => RotationState::Initial
        }
    }

    fn rotate_left(&mut self) {
        self.current_rotation = match self.current_rotation {
            RotationState::Initial => RotationState::CounterClockwise,
            RotationState::Clockwise => RotationState::Initial,
            RotationState::CounterClockwise => RotationState::Deg180,
            RotationState::Deg180 => RotationState::Clockwise
        }
    }

    pub fn get_rotation_type(&self, rotation: RotationDirection) -> (RotationMove, &[Point]) {
        match self.current_rotation {
            RotationState::Initial => match rotation {
                RotationDirection::Clockwise => (RotationMove::InitToRight, get_points_for_piece(self.piece_type, RotationState::Clockwise)),
                RotationDirection::CounterClockwise => (RotationMove::InitToLeft, get_points_for_piece(self.piece_type, RotationState::CounterClockwise)),
                RotationDirection::Deg180 => (RotationMove::InitToDeg180, get_points_for_piece(self.piece_type, RotationState::Deg180)),
            }
            RotationState::Clockwise => match rotation {
                RotationDirection::Clockwise => (RotationMove::RightToDeg180, get_points_for_piece(self.piece_type, RotationState::Deg180)),
                RotationDirection::CounterClockwise => (RotationMove::RightToInit, get_points_for_piece(self.piece_type, RotationState::Initial)),
                RotationDirection::Deg180 => (RotationMove::Deg180ToLeft, get_points_for_piece(self.piece_type, RotationState::CounterClockwise)),
            }
            RotationState::Deg180 => match rotation {
                RotationDirection::Clockwise => (RotationMove::Deg180ToLeft, get_points_for_piece(self.piece_type, RotationState::CounterClockwise)),
                RotationDirection::CounterClockwise => (RotationMove::Deg180ToRight, get_points_for_piece(self.piece_type, RotationState::Clockwise)),
                RotationDirection::Deg180 => (RotationMove::Deg180ToInit, get_points_for_piece(self.piece_type, RotationState::Initial)),
            }
            RotationState::CounterClockwise => match rotation {
                RotationDirection::Clockwise => (RotationMove::LeftToInit, get_points_for_piece(self.piece_type, RotationState::Initial)),
                RotationDirection::CounterClockwise => (RotationMove::LeftToDeg180, get_points_for_piece(self.piece_type, RotationState::Deg180)),
                RotationDirection::Deg180 => (RotationMove::InitToRight, get_points_for_piece(self.piece_type, RotationState::Clockwise)),
            }
        }
    }
}

#[cfg(test)]
mod tests {
/*    use crate::piece::rotate_array;

    #[test]
    fn rotate_matrix() {
        let mut r1 = vec![0, 0, 1];
        let mut r2 = vec![1, 1, 1];
        let mut r3 = vec![0, 0, 0];

        let mut piece_j = vec![
            &mut r1, &mut r2, &mut r3
        ];

        let mut r4 = vec![1, 1, 0];
        let mut r5 = vec![0, 1, 0];
        let mut r6 = vec![0, 1, 0];

        let rotated_piece_j = vec![
            &mut r4, &mut r5, &mut r6
        ];

        rotate_array(&mut piece_j);
        assert_eq!(piece_j, rotated_piece_j);
    }*/
}

File: csharp/Quader/Quader.Engine/Pieces/Piece.cs
using Quader.Engine.Boards;
using Quader.Engine.Pieces.WallKick;
using Quader.Engine.Primitives;

namespace Quader.Engine.Pieces;

public readonly struct WallKickCheckParams(IEnumerable<Point> tests, IEnumerable<Point> expectedPos)
{
    public readonly IEnumerable<Point> Tests = tests;
    public readonly IEnumerable<Point> ExpectedPos = expectedPos;
}

public class Piece
{
    public PieceType PieceType { get; }
    public CellType CellType { get; }
    public OffsetType OffsetType { get; }
    public Rectangle Bounds { get; private set; }

    private int _x;

    public int X
    {
        get => _x;
        set => SetX(value);
    }

    private int _y;

    public int Y
    {
        get => _y;
        set => SetY(value);
    }

    public RotationState CurrentRotation { get; private set; }
    public WallKickType WallKickType { get; }

    public Color Color => PieceHelpers.PieceTypeToColor(PieceType);

    public Piece(PieceType type)
    {
        PieceType = type;
        CurrentRotation = RotationState.Initial;
        _x = 0;
        _y = 0;

        WallKickType = PieceType switch
        {
            PieceType.I => WallKickType.PieceI,
            PieceType.O => WallKickType.PieceO,
            _ => WallKickType.Default
        };

        OffsetType = PieceHelpers.PieceTypeToOffsetType(PieceType);
        CellType = PieceHelpers.PieceTypeToCellType(PieceType);

        var initPos = GetPoints();
        Bounds = PieceHelpers.CalcBounds(initPos, X, Y);
    }

    public void Reset()
    {
        CurrentRotation = RotationState.Initial;
        Bounds = CalcBounds();
    }

    public Point[] GetPoints() => PieceHelpers.GetPointsForPiece(PieceType, CurrentRotation);
    public Point[] GetPositions() => GetPoints();

    public void SetX(int x)
    {
        _x = x;
        Bounds = CalcBounds();
    }

    public void SetY(int y)
    {
        _y = y;
        Bounds = CalcBounds();
    }

    public void MoveLeft() => X -= 1;
    public void MoveRight() => X += 1;
    public void MoveDown() => Y += 1;

    public void Rotate(RotationDirection rotation, int xOffset, int yOffset)
    {
        RotateSimple(rotation);

        var ix = X + xOffset;
        var iy = Y + yOffset;

        _x = ix;
        _y = iy;

        Bounds = CalcBounds();
    }

    public (RotationMove Move, Point[] Points) GetRotationType(RotationDirection rotation)
    {
        return CurrentRotation switch
        {
            RotationState.Initial => rotation switch
            {
                RotationDirection.Clockwise => (RotationMove.InitToRight,
                    PieceHelpers.GetPointsForPiece(PieceType, RotationState.Clockwise)),
                RotationDirection.CounterClockwise => (RotationMove.InitToLeft,
                    PieceHelpers.GetPointsForPiece(PieceType, RotationState.CounterClockwise)),
                RotationDirection.Deg180 => (RotationMove.InitToDeg180,
                    PieceHelpers.GetPointsForPiece(PieceType, RotationState.Deg180)),
                _ => throw new ArgumentOutOfRangeException(nameof(rotation), rotation, null)
            },
            RotationState.Clockwise => rotation switch
            {
                RotationDirection.Clockwise => (RotationMove.RightToDeg180,
                    PieceHelpers.GetPointsForPiece(PieceType, RotationState.Deg180)),
                RotationDirection.CounterClockwise => (RotationMove.RightToInit,
                    PieceHelpers.GetPointsForPiece(PieceType, RotationState.Initial)),
                RotationDirection.Deg180 => (RotationMove.Deg180ToLeft,
                    PieceHelpers.GetPointsForPiece(PieceType, RotationState.CounterClockwise)),
                _ => throw new ArgumentOutOfRangeException(nameof(rotation), rotation, null)
            },
            RotationState.Deg180 => rotation switch
            {
                RotationDirection.Clockwise => (RotationMove.Deg180ToLeft,
                    PieceHelpers.GetPointsForPiece(PieceType, RotationState.CounterClockwise)),
                RotationDirection.CounterClockwise => (RotationMove.Deg180ToRight,
                    PieceHelpers.GetPointsForPiece(PieceType, RotationState.Clockwise)),
                RotationDirection.Deg180 => (RotationMove.Deg180ToInit,
                    PieceHelpers.GetPointsForPiece(PieceType, RotationState.Initial)),
                _ => throw new ArgumentOutOfRangeException(nameof(rotation), rotation, null)
            },
            RotationState.CounterClockwise => rotation switch
            {
                RotationDirection.Clockwise => (RotationMove.LeftToInit,
                    PieceHelpers.GetPointsForPiece(PieceType, RotationState.Initial)),
                RotationDirection.CounterClockwise => (RotationMove.LeftToDeg180,
                    PieceHelpers.GetPointsForPiece(PieceType, RotationState.Deg180)),
                RotationDirection.Deg180 => (RotationMove.InitToRight,
                    PieceHelpers.GetPointsForPiece(PieceType, RotationState.Clockwise)),
                _ => throw new ArgumentOutOfRangeException(nameof(rotation), rotation, null)
            },
            _ => throw new ArgumentOutOfRangeException()
        };
    }

    internal void RotateSimple(RotationDirection rotation)
    {
        switch (rotation)
        {
            case RotationDirection.Clockwise:
                RotateRight();
                break;
            case RotationDirection.CounterClockwise:
                RotateLeft();
                break;
            case RotationDirection.Deg180:
                RotateRight();
                RotateRight();
                break;
            default:
                throw new ArgumentOutOfRangeException(nameof(rotation), rotation, null);
        }

        Bounds = CalcBounds();
    }

    void RotateRight()
    {
        CurrentRotation = CurrentRotation switch
        {
            RotationState.Initial => RotationState.Clockwise,
            RotationState.Clockwise => RotationState.Deg180,
            RotationState.Deg180 => RotationState.CounterClockwise,
            RotationState.CounterClockwise => RotationState.Initial,
            _ => throw new ArgumentOutOfRangeException()
        };
    }

    void RotateLeft()
    {
        CurrentRotation = CurrentRotation switch
        {
            RotationState.Initial => RotationState.CounterClockwise,
            RotationState.Clockwise => RotationState.Initial,
            RotationState.Deg180 => RotationState.Clockwise,
            RotationState.CounterClockwise => RotationState.Deg180,
            _ => throw new ArgumentOutOfRangeException()
        };
    }

    private Rectangle CalcBounds() => PieceHelpers.CalcBounds(GetPoints(), X, Y);
}